File: problems/820/solution.cpp
struct TrieNode {
    unordered_map<char, TrieNode*> children;
    bool endOfWord = false;
};

class Trie {
private:
    TrieNode *root = new TrieNode();

public:
    Trie() {}
    
    void insert(const string& s) {
        TrieNode *cur = root;
        
        for (auto it = s.rbegin(); it != s.rend(); it++) {
            const char ch = *it;
            
            if (not cur->children.count(ch)) {
                cur->children[ch] = new TrieNode();
            }
            
            cur = cur->children[ch];
        }
        
        cur->endOfWord = true;
    }
    
    int countPathsInTrie() {
        int ans = 0;
        queue<pair<TrieNode*, int>> q;
        
        q.push({root, 0});
        
        while (not q.empty()) {
            auto [node, len] = q.front();
            q.pop();
            
            len++;
            
            if (node->children.empty()) {
                ans += len;
                continue;
            }
            
            for (auto [ignore, next]: node->children) {
                q.push({next, len});
            }
        }
        
        return ans;
    }
};

class Solution {
public:
    int minimumLengthEncoding(vector<string>& words) {
        Trie *trie = new Trie();
        
        for (const string& word: words) {
            trie->insert(word);
        }
        
        return trie->countPathsInTrie();
    }
};

File: problems/820/solution.go
package main
type TrieNode struct {
    Children map[byte]*TrieNode
    EndOfWord bool
}

type Trie struct {
    root *TrieNode
}

func NewTrieNode() *TrieNode {
    return &TrieNode{
        make(map[byte]*TrieNode),
            false,
    }
}

func NewTrie() *Trie {
    return &Trie{
        NewTrieNode(),
    }
}

func (this *Trie) Insert(word string) {
    cur := this.root
    
    for i := len(word) - 1; i >= 0; i-- {
        ch := word[i]
        
        if _, ok := cur.Children[ch]; !ok {
            cur.Children[ch] = NewTrieNode()
        }
        
        cur = cur.Children[ch]
    }
    
    cur.EndOfWord = true
}

func (this *Trie) SumPathsLen() int {
    type pair struct {
        node *TrieNode
        len int
    }
    
    ans, q := 0, []pair{pair{this.root, 0}}
    
    for len(q) > 0 {
        nq := []pair{}
        
            for _, np := range q {
                np.len++

                if len(np.node.Children) == 0 {
                    ans += np.len
                    continue
                }
                
                for _, next := range np.node.Children {
                    nq = append(nq, pair{next, np.len});
                }
            }
        
        q = nq
    }
    

    
    return ans
}

func minimumLengthEncoding(words []string) int {
    trie := NewTrie()
    
    for _, word := range words {
        trie.Insert(word)
    }
    
    return trie.SumPathsLen()
}